[RFC] vim-patch: 7.4.1285

KillTheMule opened the pull request:
### vim-patch:7.4.1285

```
Problem:    Cannot measure elapsed time.
Solution:   Add reltimefloat().
```

vim/vim@79c2c88

Applied manually. None of the ifdef's applies anymore, and proftime_T was
changed into an uint64_T, so the function profile_float to convert proftime_T to
float is not needed in nvim.
### vim-patch:cb00f03

`Add missing test file.`

vim/vim@cb00f03

Converted to a lua test.
### Add documentation for reltimefloat()

from vim/vim@03413f4

Also adjust  the entries for reltime() and reltimestr().

KillTheMule committed:
vim-patch:7.4.1285

Problem:    Cannot measure elapsed time.
Solution:   Add reltimefloat().

https://github.com/vim/vim/commit/79c2c881bb7ae1cbdeeff91d4875b4bf2e54df06

Applied manually. None of the ifdef's applies anymore, and proftime_T was
changed into an uint64_T, so the function profile_float to convert proftime_T to
float is not needed in nvim.

KillTheMule committed:
vim-patch:cb00f03

Add missing test file.

https://github.com/vim/vim/commit/cb00f039332d3188931035e9d07144546fdea78a

Converted to a lua test. Change the tolerance of the test to avoid false
positives on travis.

KillTheMule committed:
Add documentation for reltimefloat()

from https://github.com/vim/vim/commit/03413f44167c4b5cd0012def9bb331e2518c83cf

Also adjust  the entries for reltime() and reltimestr().

KillTheMule commented:
Default built failure seems unrelated, but travis fails on the new test, but just for clang on osx... looks like it took quite some time, and part of is is sleeping for 10ms and testing for that time difference. Could this just be a slow environment?

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -2007,6 +2007,7 @@ range( {expr} [, {max} [, {stride}]])
 readfile( {fname} [, {binary} [, {max}]])
 				List	get list of lines from file {fname}
 reltime( [{start} [, {end}]])	List	get time value
+reltimefloat({time})		Float	turn the time value into a Float
```
The space after the paren in this list of functions is intentional so that `ctrl-]` works. 

    reltimefloat( {time})

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -5303,7 +5314,6 @@ reltimestr({time})				*reltimestr()*
 		can use split() to remove it. >
 			echo split(reltimestr(reltime(start)))[0]
 <		Also see |profiling|.
-		{only available when compiled with the |+reltime| feature}
```
The note `The accuracy depends on the system.` can be removed also (line 5301/5312). Nvim's `reltimestr()` should be very accurate on all supported systems. (We removed use of `gettimeofday()` in db7cd61f6297aab6343f7c6cd424d8987ca03a58)

KillTheMule replied:
I removed the line, was on 5340 though.

justinmk commented on test/functional/eval/reltime_spec.lua:
```diff
@@ -0,0 +1,36 @@
+local neq, execute, funcs  = helpers.neq, helpers.execute, helpers.funcs
+local reltime, reltimestr, reltimefloat = funcs.reltime, funcs.reltimestr, funcs.reltimefloat
+
+describe('Basic functionality of reltimestr and reltimefloat', function()
```
describe('reltimestr(), reltimefloat()', function()

Try to be brief yet precise in `describe()` and each nested `it()`: the titles are concatenated to make a single line. And it's a good idea in general.

justinmk commented on test/functional/eval/reltime_spec.lua:
```diff
@@ -0,0 +1,36 @@
+
+    neq(reltimestr(elapsed), '0.0')
+    print(reltimestr(elapsed))
+    ok((string.match(reltimestr(elapsed), "0%.0") and true) or false)
```
Usually we test `~= nil`.

    ok(nil ~= string.match(reltimestr(elapsed), "0%.0"))

KillTheMule commented:
Changed stuff and rebased. Indeed brevity in names and descriptions is a real weakness of mine :D  The note about system-dependency of the time was at a totally different line for me, but I guess you weren't looking at my branch.

KillTheMule commented:
Travis failure is only gcc on osx, and the test took quite some time, so it's reasonable to assume that 'sleep(10m)' is less deterministic than hoped for. I'd hate to introduce such a flaky test, should I think about a different one?

KillTheMule commented:
I increased the tolerance of the test and fixed another small oversight.

KillTheMule commented:
QB failures should be fixed by https://github.com/neovim/neovim/pull/4718.

justinmk commented:
Verified test. Thank @KillTheMule
